### DISCUSSION TITLE
[Errno 32] Pipe Error using PyUSB; SLM-25 Decibel Meter Connected to Laptop

### DISCUSSION THREAD

**Discussion by 808reefer** [pyusb/pyusb#411] [Q&A] [answered]
Hi,

I am having trouble trying to obtain data from a SLM-25 sound level meter from Gain Express using PyUSB. Laptop OS is Windows 10, PyUSB version 1.2.1, and Python version 3.9.10. I have used Zadig to replace the drivers with WinUSB. It seems like there is a Pipe error, but not sure how to fix it. I have been stuck for 3 weeks and am very desperate. Any help would be greatly appreciated, thank you! I have the Code, Output, and Error posted below:

Code:

```py
#!/usr/bin/python

import sys
import usb.core
import usb.util
import time
import unittest
import usb.core
import usb._interop
from usb._debug import methodtrace
import usb.util
import usb.backend.libusb0 as libusb0
import usb.backend.libusb1 as libusb1
import usb.backend.openusb as openusb

dev=usb.core.find(idVendor=0x10c4,idProduct=0x82cd)

assert dev is not None

print (dev)

print (hex(dev.idVendor)+','+hex(dev.idProduct))

dev.set_configuration()

while True:
    time.sleep(1)
    try:
        #ret = dev.read(0x81, 200, 100)  #results in USBTimeoutError
        ret = dev.ctrl_transfer(bmRequestType=0xC0,bRequest=0x04,wValue=0,wIndex=0,data_or_wLength=200)
        dB = (ret[0]+((ret[1]&3)*256))*0.1+30
        print (dB)
    finally:
        print("success")
```

Output:

```
Device:
DEVICE ID 10c4:82cd on Bus 001 Address 005 =================
 bLength                :   0x12 (18 bytes)
 bDescriptorType        :    0x1 Device
 bcdUSB                 :  0x110 USB 1.1
 bDeviceClass           :    0x0 Specified at interface
 bDeviceSubClass        :    0x0
 bDeviceProtocol        :    0x0
 bMaxPacketSize0        :   0x40 (64 bytes)
 idVendor               : 0x10c4
 idProduct              : 0x82cd
 bcdDevice              :    0x0 Device 0.0
 iManufacturer          :    0x1 SLAB
 iProduct               :    0x2 WL100
 iSerialNumber          :    0x0
 bNumConfigurations     :    0x1
  CONFIGURATION 1: 64 mA ===================================
   bLength              :    0x9 (9 bytes)
   bDescriptorType      :    0x2 Configuration
   wTotalLength         :   0x29 (41 bytes)
   bNumInterfaces       :    0x1
   bConfigurationValue  :    0x1
   iConfiguration       :    0x0
   bmAttributes         :   0x80 Bus Powered
   bMaxPower            :   0x20 (64 mA)
    INTERFACE 0: Human Interface Device ====================
     bLength            :    0x9 (9 bytes)
     bDescriptorType    :    0x4 Interface
     bInterfaceNumber   :    0x0
     bAlternateSetting  :    0x0
     bNumEndpoints      :    0x2
     bInterfaceClass    :    0x3 Human Interface Device
     bInterfaceSubClass :    0x0
     bInterfaceProtocol :    0x0
     iInterface         :    0x0
      ENDPOINT 0x81: Interrupt IN ==========================
       bLength          :    0x7 (7 bytes)
       bDescriptorType  :    0x5 Endpoint
       bEndpointAddress :   0x81 IN
       bmAttributes     :    0x3 Interrupt
       wMaxPacketSize   :   0x40 (64 bytes)
       bInterval        :    0xa
      ENDPOINT 0x1: Interrupt OUT ==========================
       bLength          :    0x7 (7 bytes)
       bDescriptorType  :    0x5 Endpoint
       bEndpointAddress :    0x1 OUT
       bmAttributes     :    0x3 Interrupt
       wMaxPacketSize   :   0x40 (64 bytes)
       bInterval        :    0xa
```

Error:

```
Exception has occurred: USBError x
[Errno 32] Pipe error
```

The error occurs at the line

```py
ret = dev.ctrl_transfer(bmRequestType=0xC0, bRequest=0x04, wValue=0, wIndex=0, data_or_wLength=200)
```

ANY HELP WILL BE APPRECIATED! MAHALOS!

**Comment by jonasmalacofilho** [comment, marked as answer]
One way to look at a problem like this is to go to the backend documentation for the call and the error.  You haven't mentioned, but I'm going to assume for now that the backend is libusb 1.0.

Under that assumption, the error is [`LIBUSB_ERROR_PIPE`](https://libusb.sourceforge.io/api-1.0/group__libusb__misc.html#gab2323aa0f04bc22038e7e1740b2f29ef), and [`libusb_control_transfer()`](https://libusb.sourceforge.io/api-1.0/group__libusb__syncio.html#gadb11f7a761bd12fc77a07f4568d56f38) is documented to return it "if the control request was not supported by the device."

That said, another thing to note is the `dev.set_configuration()` call, and the commented out line:

```py
#ret = dev.read(0x81, 200, 100)  #results in USBTimeoutError
```

This might be relevant because `dev.set_configuration()` can act as a soft-reset in some cases, and that could explain the mentioned timeout.  More information about this can be found in out [FAQ](https://github.com/pyusb/pyusb/blob/master/docs/faq.rst#how-not-to-call-set_configuration-on-a-device-already-configured-with-the-selected-configuration), as well as in the libusb FAQ that ours links to.

---

P.S. In the future, posting the complete exception and backtrace is preferable, as there might be something there that is meaningful to me or others.  And, when debugging errors, you should also run your example with `PYUSB_DEBUG=debug` and `LIBUSB_DEBUG=4` set on your environment, as that can help track what's going on at the lower levels.

P.P.S. Moving this to the Discussions tab, Q/A.

**Comment by jpswensen** [comment]
@808reefer , did you ever get the USB reading of the decibel working from python?
